### PR TITLE
error log handle to avoid log flooding when remote endpoints not working

### PR DESCRIPTION
**Pull request by marcolan018**
https://github.com/stolostron/backlog/issues/20435

Signed-off-by: Marco llan@redhat.com

### CODE REVIEW

File: internal/remotewrite/proxy.go [hunk 1]
```diff
@@ -57,41 +59,72 @@ func remoteWrite(write *url.URL, endpoints []Endpoint, logger log.Logger) http.H
...
+									messageKey: "Failed to forward metrics" + resp.Status,
+									keyvals: []interface{}{
+										"msg", "Failed to forward metrics", "return code", resp.Status, "url", ep.URL,
+									}}
 							} else {
-								level.Error(rlogger).Log("msg", "Failed to forward metrics", "return code", resp.Status, "response", string(responseBody), "url", ep.URL)
+								LogChannels[j] <- logMessage{
+									messageKey: "Failed to forward metrics" + resp.Status,
+									keyvals: []interface{}{
+										"msg", "Failed to forward metrics", "return code", resp.Status, "response", string(responseBody), "url", ep.URL}}
```
Comment (clyang82) [inline internal/remotewrite/proxy.go:121]: What is purpose to handle `err != nil` and `err == nil` separately?

File: internal/remotewrite/proxy.go [hunk 2]
```diff
@@ -57,41 +59,72 @@ func remoteWrite(write *url.URL, endpoints []Endpoint, logger log.Logger) http.H
...
 				}
 			}
 
 			req, err := http.NewRequest(http.MethodPost, endpoint.URL, bytes.NewReader(body))
 			req.Header = r.Header
 			if err != nil {
-				level.Error(rlogger).Log("msg", "Failed to create the forward request", "err", err, "url", endpoint.URL)
+				//level.Error(rlogger).Log("msg", "Failed to create the forward request", "err", err, "url", endpoint.URL)
+				LogChannels[i] <- logMessage{
+					messageKey: "Failed to create the forward request",
```
Comment (clyang82) [inline internal/remotewrite/proxy.go:88]: Suggest to use lower case for all messages to comply with the golang standard.

File: internal/remotewrite/logchannel.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		if time.Since(*counter.logStartTime) >= logInterval {
+			message := fmt.Sprintf("Error occurred %d times in last %d seconds: %s",
+				counter.logNumber, int(time.Since(*counter.logStartTime).Seconds()), key)
+			keyvals[1] = message
+			level.Error(rlogger).Log(keyvals...)
+			if counter.logNumber < logMaxCount {
+				counter.reducedLog = false
+				counter.LogTimestamps = []time.Time{}
+			} else {
+				counter.logNumber = 0
```
Comment (clyang82) [inline internal/remotewrite/logchannel.go:101]: Why start with 0 for logNumber? Can you add some test cases to help me understand the logic here?

**Review by clyang82 — CHANGES_REQUESTED** [review]

File: internal/remotewrite/proxy.go (hunk 1 above)
Comment (marcolan018) [reply internal/remotewrite/proxy.go:121]: err == nil means the response body can be parsed, then we can log it out for better assist in PD

File: internal/remotewrite/proxy.go (hunk 2 above)
Comment (marcolan018) [reply internal/remotewrite/proxy.go]: done

File: internal/remotewrite/logchannel.go (hunk 3 above)
Comment (marcolan018) [reply internal/remotewrite/logchannel.go:109]: done

**Pull request merged**